File: foodOrderingSystem/go-without-multithreading/strategy/highest_rating_strategy.go
package strategy
import "app/foodOrderingSystem/go-without-multithreading/entities"

type HighestRatingStrategy struct{}

func (s *HighestRatingStrategy) SelectRestaurant(eligibleRestaurants []*entities.Restaurant, requestedItems map[string]int) *entities.Restaurant {
	var bestRestaurant *entities.Restaurant
	maxRating := -1.0

	for _, r := range eligibleRestaurants {
		if r.Rating > maxRating {
			maxRating = r.Rating
			bestRestaurant = r
		}
	}
	return bestRestaurant
}

File: foodOrderingSystem/java-without-multithreading/strategy/HighestRatingStrategy.java
package strategy;
import entities.Restaurant;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class HighestRatingStrategy implements RestaurantSelectionStrategy {
    @Override
    public Optional<Restaurant> selectRestaurant(List<Restaurant> eligibleRestaurants, Map<String, Integer> requestedItems) {
        Restaurant bestRestaurant = null;
        double maxRating = -1.0;

        for (Restaurant r : eligibleRestaurants) {
            if (r.getRating() > maxRating) {
                maxRating = r.getRating();
                bestRestaurant = r;
            }
        }
        return Optional.ofNullable(bestRestaurant);
    }
}
